pkg/crd/osm/v1/operating_system_profile.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2021 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	// OperatingSystemProfileResourceName represents "Resource" defined in Kubernetes
+	OperatingSystemProfileResourceName = "operatingsystemprofile"
+
+	// OperatingSystemProfileKindName represents "Kind" defined in Kubernetes
+	OperatingSystemProfileKindName = "OperatingSystemProfile"
+)
+
+//+genclient
+//+genclient:Namespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// OperatingSystemProfile is the object that represents the OperatingSystemProfile
+type OperatingSystemProfile struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// OperatingSystemProfileSpec represents the operating system confuration spec.
+	Spec OperatingSystemProfileSpec `json:"spec"`
+}
+
+// OperatingSystemProfileSpec represents the data in the newly created OperatingSystemProfile
+type OperatingSystemProfileSpec struct {
+	// OSType represent the operating system name e.g: ubuntu
+	OSName string `json:"osName"`
+	// OSVersion the version of the operaitng system
+	OSVersion string `json:"osVersion"`
+	// SupportedCloudProviders represent the cloud providers that support the os name and version
+	SupportedCloudProviders []SupportedCloudProvider `json:"supportedCloudProviders"`
+	// Units a list of units that represent the systemd units which will run on the instance
+	Units []Unit `json:"units,omitempty"`
+	// Files a list of files that should exist in the instance
+	Files []File `json:"files,omitempty"`
+}
+
+// SupportedCloudProvider
+type SupportedCloudProvider struct {
+	// Name represents the name of the supported cloud provider
+	Name string `json:"name"`
+	// Spec represents the os/image reference in the supported cloud provider
+	Spec runtime.RawExtension `json:"spec"`
+}
+
+// Unit is a systemd unit used for the operating system config.
+type Unit struct {
+	// Name is the name of a unit.
+	Name string `json:"name"`
+	// Command is the unit's command.
+	Command *string `json:"command,omitempty"`
+	// Enable describes whether the unit is enabled or not.
+	Enable *bool `json:"enable,omitempty"`
+	// Content is the unit's content.
+	Content *string `json:"content,omitempty"`
```

Comment:
What's exactly supposed to be represented with the `Content` field? To my understanding, this should be the content of the actual unit file. If that's the case, then why do we need the `Command` field, since the command will be present in the unit file (i.e. in the `Content` field)?

Reply:
I actually added this field in order to change how we call those scripts and if they change during the lifecycle of the machine, e.g: instead of changing the file explicitly in order to choose how to run this unit, we change the command and the controller would take care of the script adaptation. I wanted to have it as an enum, where we can specify how the script will be run. But you are right at this point this might complicate things. I will remove it for now and if we think of making the OSM keep observing the state of the machines this field is vital